src/rust/cryptography-x509/src/certificate.rs

```diff
@@ -13,6 +13,21 @@ pub struct Certificate<'a> {
     pub signature: asn1::BitString<'a>,
 }
 
+impl Certificate<'_> {
+    /// Retrieves the extension identified by the given OID,
+    /// or None if the extension is not present (or no extensions are present).
+    pub fn get_extension(&self, oid: asn1::ObjectIdentifier) -> Option<extensions::Extension> {
```

Comment:
One thing to note: currently the requirement that OIDs not be repeated is enforced in the python bindings layer, but we want to be sure we're enforcing that, else you can get clowny behavior (if there's two extensions with the same OID, which wins?)

Reply:
Yeah, I was thinking about solutions to that -- it'd be nice if we could dedupe and do the extension uniqueness check exactly once (in the Rust layer).

Two random ideas:

1. Lazy: add a `Certificate::check_invariants()` (better name needed) or similar API, which users *must* invoke to correctly iterate over extensions, etc.
2. More involved: make `Certificate` correct by construction by having it be a newtype of `Certificate(ParsedCertificate)`, where `ParsedCertificate` is the current `CertificateType`. From there, the only allowed construction of `Certificate` would be via a `TryFrom`.

Thoughts?

Reply:
Because reasons we currently only check for duplicate extensions when accessing `.extensions`, not on initial load.

I'd probably start with a `check_duplicate_extensions()` method that can be called from both places.

Reply:
Sounds good; that also gives me transitive coverage 🙂